src/osd/osd_types.h

```diff
@@ -2480,6 +2480,8 @@ struct pg_info_t {
   pg_history_t history;
   pg_hit_set_history_t hit_set;
 
+  bool have_missing;
```

Comment:
should add this to ```operator==``` and ```operator<<```for completeness.

```encode()``` and ```decode()``` also need this added so it is actually sent across the wire

Reply:
```have_missing``` would only be potentially true for ```pg_info_t``` sent to the primary, and stored in ```peer_info```. It's only used before the next step of peering when we fetch the actual log and missing set from the secondaries. Once we have actual missing set, ```peer_info[peer]->have_missing``` shouldn't be read again.

These ```peer_info``` will be reset whenever peering restarts, and I think all that is needed is to make ```PG::proc_replica_info()``` update ```peer_info[peer]->have_missing``` even when ```last_update``` matches the  ```pg_info_t``` we already have for that peer.

Reply:
thanks the guidance, updated